### PR TITLE
updated psl; no cast needed anymore

### PR DESCRIPTION
**Pull request by Maetes**
Please fill in this template.

- [X] Use a meaningful title for the pull request. Include the name of the package modified.
- [X] Test the change in your own code. (Compile and run.)
- [X] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [X] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [X] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [X] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If changing an existing definition:
- [X] Provide a URL to documentation or source code which provides context for the suggested changes: https://ibb.co/CbZq0vx
- [X] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.

### CODE REVIEW

**Review by BendingBender — CHANGES_REQUESTED** [review]
I'm not sure what you were trying to achieve but for me it looks like that you were trying to prevent explicit casts for the result of the `parse` function. This is actually unnecessary if you guard for the `error` property:

```ts
import { parse } from "psl";

const result = parse("foo");

if (result.error) {
    const error = result;
} else {
    const res = result;
}
```

You can see in this example that you don't need to cast the result:

type of error

type of result

File: types/psl/index.d.ts [hunk 1]
```diff
@@ -1,39 +1,39 @@
...
 export interface ParsedDomain {
-    tld: string | null;
-    sld: string | null;
-    domain: string | null;
-    subdomain: string | null;
-    listed: boolean;
+    tld?: string;
+    sld?: string;
+    domain?: string;
+    subdomain?: string;
```
[the remark this answers is not in the document]
Comment (Maetes) [reply types/psl/index.d.ts:21]: It returns null, yes. Just wanted to make it shorter. Regarding this example. Does it make a difference?

File: types/psl/index.d.ts [hunk 2]
```diff
@@ -1,39 +1,39 @@
...
+    listed?: boolean;
```
[the remark this answers is not in the document]
Comment (Maetes) [reply types/psl/index.d.ts:22]: True, keeping boolean is correct.

File: types/psl/index.d.ts [hunk 3]
```diff
@@ -1,39 +1,39 @@
...
     input: string;
-    error: undefined;
 }
 
 export interface ParseError {
     input: string;
-    error: {
+    error?: {
```
[the remark this answers is not in the document]
Comment (Maetes) [reply types/psl/index.d.ts:28]: Of course you are right. This is a mistake.

File: types/psl/index.d.ts [hunk 4]
```diff
@@ -1,39 +1,39 @@
...
-// Definitions by: BendingBender <https://github.com/BendingBender>
+// Definitions by:  BendingBender <https://github.com/BendingBender>
+//                  Maetes <https://github.com/Maetes>
 // Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
 // TypeScript Version: 2.1
 
 export as namespace psl;
 
-export function parse(domain: string): ParsedDomain | ParseError;
+export function parse(domain: string): ParsedDomain & ParseError;
```
[the remark this answers is not in the document]
Comment (Maetes) [reply types/psl/index.d.ts:10]: Overall i tried to achieve to not cast, right. When i see your screenshots its totally clear for me and i wouldn't had create anything... Maybe i missed something in my config, cause in my project im not able to get to the right Type by using a typeguard. I'll retry and say sorry for that, maybe it was too early for me today ;)

File: types/psl/index.d.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (BendingBender) [reply types/psl/index.d.ts:21]: With `strictNullChecks` it actually makes a difference, testing for `null` is not the same as testing for `undefined`. Practically, it makes almost no difference but in some cases it does: default parameters for functions won't choose the default value when provided a `null`, for example. And the `===` operator differentiates between `null` and `undefined`. And there are more edge cases like this.

File: types/psl/index.d.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (BendingBender) [reply types/psl/index.d.ts:10]: You should maybe upgrade your typescript compiler, older versions are not very good in type inference.

**Pull request closed without merging**